// @flagship.io/react-sdk/dist/FlagshipErrorBoundary.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var ReactErrorBoundaryContainer_1 = __importDefault(require("./components/ReactErrorBoundaryContainer"));
var FlagshipErrorBoundary = (function (_super) {
    __extends(FlagshipErrorBoundary, _super);
    function FlagshipErrorBoundary(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            error: _this.props.error,
            errorInfo: null,
            isCollapse: false
        };
        return _this;
    }
    FlagshipErrorBoundary.prototype.componentDidCatch = function (error, errorInfo) {
        var isSafeModeEnabled = this.props.sdkSettings.enableSafeMode;
        if (isSafeModeEnabled) {
            this.props.log.fatal("An error occurred. The SDK is switching into safe mode:\n" + error.stack);
            this.setState({
                error: error,
                errorInfo: errorInfo
            });
        }
    };
    FlagshipErrorBoundary.getDerivedStateFromError = function (error) {
        return {
            error: error
        };
    };
    FlagshipErrorBoundary.prototype.componentDidUpdate = function (prevProps, prevState) {
        var isSafeModeEnabled = this.props.sdkSettings.enableSafeMode;
        if (isSafeModeEnabled) {
            if (this.state.error !== prevState.error) {
                this.props.onError(this.state.error);
            }
            if (this.props.error !== prevProps.error) {
                this.props.log.fatal("An error occurred. The SDK is switching into safe mode:\n" + this.props.error.stack);
            }
        }
    };
    FlagshipErrorBoundary.prototype.render = function () {
        var _this = this;
        var _a = this.state, error = _a.error, isCollapse = _a.isCollapse;
        var _b = this.props, children = _b.children, handleDisplay = _b.handleDisplay, customerChildren = _b.customerChildren, errorProp = _b.error, _c = _b.sdkSettings, nodeEnv = _c.nodeEnv, enableErrorLayout = _c.enableErrorLayout, enableSafeMode = _c.enableSafeMode;
        if (enableSafeMode && (errorProp || error)) {
            var errorCopy = errorProp || error;
            if (handleDisplay) {
                return handleDisplay({
                    children: children,
                    debugMode: nodeEnv !== 'production' && !!enableErrorLayout,
                    isCollapse: isCollapse,
                    error: errorCopy,
                    onClickCollapse: function () {
                        _this.setState(__assign(__assign({}, _this.state), { isCollapse: !isCollapse }));
                    }
                });
            }
            return (react_1.default.createElement(ReactErrorBoundaryContainer_1.default, { debugMode: nodeEnv !== 'production' && !!enableErrorLayout, isCollapse: isCollapse, error: errorCopy, onClickCollapse: function () {
                    _this.setState(__assign(__assign({}, _this.state), { isCollapse: !isCollapse }));
                } }, customerChildren));
        }
        return children;
    };
    return FlagshipErrorBoundary;
}(react_1.default.Component));
exports.default = FlagshipErrorBoundary;
