// @flagship.io/react-sdk/dist/tests/mock/apiAnswers/oneModifInMoreThanOneCampaign.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var env_1 = require("../env");
var oneModifInMoreThanOneCampaign = {
    visitorId: env_1.vId,
    campaigns: [
        {
            id: 'blntcamqmdvg04g371f0',
            variationGroupId: 'blntcamqmdvg04g371h0',
            variation: {
                id: 'blntcamqmdvg04g371hg',
                modifications: {
                    type: 'FLAG',
                    value: {
                        psp: 'dalenys'
                    }
                }
            }
        },
        {
            id: 'bmjdprsjan0g01uq2crg',
            variationGroupId: 'bmjdprsjan0g01uq2csg',
            variation: {
                id: 'bmjdprsjan0g01uq2ctg',
                modifications: {
                    type: 'JSON',
                    value: {
                        algorithmVersion: 'new'
                    }
                }
            }
        },
        {
            id: 'bmjdprsjan0g01uq2ceg',
            variationGroupId: 'bmjdprsjan0g01uq2ceg',
            variation: {
                id: 'bmjdprsjan0g01uq1ctg',
                modifications: {
                    type: 'JSON',
                    value: {
                        algorithmVersion: 'yolo2',
                        psp: 'yolo',
                        hello: 'world'
                    }
                }
            }
        }
    ]
};
exports.default = oneModifInMoreThanOneCampaign;
