// @flagship.io/react-sdk/dist/components/ReactErrorBoundaryContainer/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var ReactErrorBoundaryContainer = function (_a) {
    var children = _a.children, debugMode = _a.debugMode, isCollapse = _a.isCollapse, onClickCollapse = _a.onClickCollapse, error = _a.error;
    var btnStyle = {
        display: 'inline-block',
        verticalAlign: 'middle',
        border: '1px solid transparent',
        padding: '0.375rem 0.75rem',
        fontSize: '1rem',
        lineHeight: 1.5,
        borderRadius: '0.25rem',
        marginLeft: '16px',
        transition: "background-color 0.15s ease-in-out,\n        border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out"
    };
    return (react_1.default.createElement(react_1.default.Fragment, null,
        debugMode && (react_1.default.createElement("div", { className: "fsErrorDebugContainer", style: {
                backgroundColor: 'red',
                minHeight: '6vh',
                position: 'fixed',
                zIndex: 9999,
                bottom: 0,
                width: '100%',
                opacity: isCollapse ? 1 : 0.4,
                display: 'flex',
                flexDirection: 'column',
                padding: '0px 16px',
                transition: 'opacity 0.5s ease-in-out'
            } },
            react_1.default.createElement("div", { style: {
                    display: 'flex',
                    alignItems: 'center',
                    width: '100%',
                    marginBottom: '8px',
                    marginTop: '8px'
                } },
                react_1.default.createElement("h3", { style: { color: 'white' } }, "Flagship React SDK has crashed."),
                react_1.default.createElement("button", { style: __assign({}, btnStyle), onClick: onClickCollapse }, isCollapse ? 'Hide details' : 'Show details')),
            error && (react_1.default.createElement("div", { style: {
                    maxHeight: isCollapse ? 200 : 0,
                    overflow: 'auto',
                    color: 'wheat',
                    whiteSpace: 'pre-line',
                    transition: 'max-height 0.5s ease-in-out'
                } }, "" + error.stack)))),
        react_1.default.createElement("div", { id: "flagshipSafeModeContainer" }, children)));
};
ReactErrorBoundaryContainer.defaultProps = {};
exports.default = ReactErrorBoundaryContainer;
